torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -3489,6 +3489,160 @@ Tensor i1e_backward(
   });
 }
 
+// lu_solve is a map (LU, B) -> (PLU)^{-1} B,
+// where LU = L + U - I and P is a permutation matrix.
+// Note that the map is not differentiable wrt P!
+//
+// Let 1 = ones_like(LU),
+// 1_U = 1.triu(),
+// 1_L = 1 - 1_U (note the zero diagonal),
+// * := the Hadamard (element-wise) product
+//
+// Forward AD:
+//
+// Let X := U^{-1} L^{-1} P^T B be the output of the function.
+// Also, the LU input of the function could be represented as
+// LU = L + U - I.
+//
+// Differentiating LU = L + U - I produces:
+// dLU = dL + dU.
+// Noting that dL and dU are lower- and upper-triangular, respectively,
+// and that the diagonal of L is never explicitly exposed, so
+// diag(dL) = 0, it follows
+// dL = dLU * 1_L,
+// dU = dLU * 1_U.
+//
+// Differentiating X = U^{-1} L^{-1} P^T B produces:
+// dX = dU^{-1} L^{-1} P^T B + U^{-1} dL^{-1} P^T B + U^{-1} L^{-1} P^T dB
+// Note that for any invertible matrix A we have A A^{-1} = I, hence
+// dA A^{-1} + A dA^{-1} = 0 => dA^{-1} = -A^{-1} dA A^{-1}.
+// Inserting it back into the definition of dX gives:
+// dX = -U^{-1} dU U^{-1} L^{-1} P^T B - U^{-1} L^{-1} dL L^{-1} P^T B + U^{-1} L^{-1} P^T dB
+//
+// Backward AD:
+//
+// Using the definition of dL, dU from above:
+// Tr(L_grad^H dL) + Tr(U_grad^H dU) = Tr(L_grad^H (dLU * 1_L)) + Tr(U_grad^H (dLU * 1_U))
+//                                   = [using Tr(A (B * C)) = Tr((A * B^T) C)
+//                                   = Tr((L_grad^H * 1_L^T) dLU) + Tr((U_grad^H * 1_U^T) dLU),
+// hence
+// LU_grad = L_grad * 1_L + U_grad * 1_U (!!!)
```

Comment:
Not a direct sum, note `tril` and `triu` being used here.

Reply:
Yeah, that's what I meant with "the same for the adjoints". This happens in general whenever we have a constrained output, only that in this case the constraint is particularly simple because it's the direct sum of two vector spaces: the upper triangular matrices and the lower triangular matrices with ones in the diagonal. As you proved, the tangent spaces of these spaces are the upper triangular matrices and the lower triangular matrices with zeros in the diagonal. The topological fact that the output forms a product of two manifolds gives this split in the gradients.

For example, same happens in `eigh` where the output is an orthogonal matrix, and then we project the gradient onto the tangent space of the orthogonal matrices at the identity: the skew-symmetric matrices.

This was more of a comment to keep in mind, feel free to mark as resolved.

Reply:
I totally agree, the point was about what to keep in the comments. I will keep as is, since only a basic algrebra is required to understand the result.